### ISSUE TITLE
Starts with big error, then fails to fetch new feeds.

### ISSUE THREAD

**Issue by impaktor** [skeeto/elfeed#283]
When starting elfeed, I get a very large error in the mini-buffer. Setting `(setq debug-on-error t)`, I get a 2 Mb error buffer.

However, after the error, I could fetch new feeds, and my old index was preserved, and everything worked like normal, so it was just an annoyance (starting some weeks ago?) when starting elfeed.

I updated to latest elfeed in Melpa today, but now, I will not allow me to fetch new feeds, nor show me my old one.

![2018-06-07-201841_957x391_scrot](https://user-images.githubusercontent.com/619390/41369422-16957a5e-6f45-11e8-841c-1c34e1e61dd8.png)

Error as of today:

``` emacs-lisp
elfeed-meta--plist: Don’t know how to access metadata on [cl-struct-elfeed-feed "http://planet.emacsen.org/atom.xml" "http://planet.emacsen.org/atom.xml" "Planet Emacsen" "" (:last-modified "Wed, 13 Jun 2018 09:04:04 GMT" :etag "\"5538d-30f09-56e8243b08900\"" :failures 34)]
```

I'm on Arch Linux, GNU Emacs 26.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.22.30) of 2018-05-29, and elfeed 20180612.1634 from MELPA

EDIT: before today, my elfeed version was maybe +1 month old, or more. Not sure how often I've updated it.

**Comment by skeeto** [comment]
Were you running a pre-release build of Emacs at some point? That first 
line in the error message is suspicious. Notice the :dummyroot keyword. 
That's the name of the first field in an avl-tree struct, but when 
printed it's not supposed to look like a plist. This suggests to me that 
your index was saved under some Emacs build where this was the case.

The index format is tied the printer of whatever Emacs wrote it. This 
wasn't a problem until Emacs 26, where structs suddenly print 
differently due to the introduction of records. Elfeed includes a 
routine for converting an index from Emacs 25 to Emacs 26, but it's only 
reliable for actual point releases.

I poked around the Emacs Git repository to find a release that printed 
like this, but didn't find anything. As far as I can tell, it doesn't 
look like Arch distributed a pre-release build.

The easiest thing to do is toss your .elfeed/ directory and start fresh. 
It *may* be possible to repair it by modifying the file by hand to set 
the format version to "0.0.3" (if it's not already), then eliminate the 
record property names. The latter can be done crudely with sed:

sed -i 's/:dummyroot\>//g' index
sed -i 's/:cmpfun\>//g' index

**Comment by impaktor** [comment]
Hi!

The TL;DR, see bottom line, your fix worked, so I'm a happy camper!

> Were you running a pre-release build of Emacs at some point? That first line in the error message is suspicious.

No. I've only been on the official arch releases since I started using elfeed. I can drill a bit further into the details of what I've been up to:

1. I run elfeed on an arch linux box, with version `GNU Emacs 25.3.1 (x86_64-pc-linux-gnu, GTK+ Version 3.22.26) of 2018-02-09` all working fine.

2. I moved my `~/.elfeed` to another arch machine 2018-05-29. Elfeed worked fine with the move, both Emacs-es were pre-26, arch build.

3. After some time (days?), Elfeed started burping a very large `[[[[[[...[[[` data structure in the minibuffer. I turned on debug logging, and [here's](http://home.thep.lu.se/~karlf/error_log_elfeed) a log (my Emacs crashes trying to parse the full error-log file). (It's timestamped 7th june).

4. As mentioned in my first post, after this error in the minibuffer, elfeed worked fine when pressing `G` for fetching new data (and it found it's index again with old feeds). Thus, I delayed reporting the error.

5. My log shows, that 2018-06-03 (last Sunday) i ran `pacman -Syu` and that apparently upgraded me to Emacs 26.1-1 in arch linux, from Emacs 25.3-4. I hadn't noticed this actually. I wonder if this is what caused the initial problem, and if I delayed 4 days with getting the debug printout mentioned above? Not impossible, but it's also possible the error started before, and was unchanged by the Emacs upgrade. Thus, I'm uncertain what the chronological order of this paragraph relative to the first occurrence of the error is. By the way, I have an `~/.elfeed/index.backup` from 2018-06-04.

6. A few days later, (Wednesday? 2018-06-13), I started emacs, upgraded elfeed through MELPA, and then started / loaded Elfeed. I'm semi-certain that I noticed that there was _no error_ at start up anymore(!), so I though "neat, it's been fixed in by the elfeed upgrade!". But alas, the next time I started elfeed it started this new misbehaving i.e. it can't load old index, nor fetch new feeds.

So, to me, it seems like it was something in how elfeed did the first save after it had been updated that completely broke the index file (if interested it's [here](http://home.thep.lu.se/~karlf/index)), but the error occurred before upgrading elfeed and maybe also before moving to emacs 26. The Emacs upgrade from 25 to 26 seems to have happened without me noticing, so I'm not 100% sure when this error started popping up relative to Emacs upgrade.

> The index format is tied the printer of whatever Emacs wrote it. This wasn't a problem until Emacs 26, where structs suddenly print differently due to the introduction of records. Elfeed includes a routine for converting an index from Emacs 25 to Emacs 26, but it's only reliable for actual point releases. I poked around the Emacs Git repository to find a release that printed like this, but didn't find anything.

> As far as I can tell, it doesn't look like Arch distributed a pre-release build.

That's weird.
``` shell
[~]% ls -l /var/cache/pacman/pkg/emacs-2*
-rw-r--r-- 1 root root 39035340 Feb  9 00:21 /var/cache/pacman/pkg/emacs-25.3-3-x86_64.pkg.tar.xz
-rw-r--r-- 1 root root 39033916 May  9 12:41 /var/cache/pacman/pkg/emacs-25.3-4-x86_64.pkg.tar.xz
-rw-r--r-- 1 root root 40104504 May 29 06:10 /var/cache/pacman/pkg/emacs-26.1-1-x86_64.pkg.tar.xz
```
Caveat: this is the computer I moved elfeed to, but elfeed initially worked
fine on this computer.

> The easiest thing to do is toss your .elfeed/ directory and start fresh.

I would really hate to loose my index, so if you can come up with something that fixes this.... I'd greatly appreciate it.

> It *may* be possible to repair it by modifying the file by hand to set the format version to "0.0.3" (if it's not already), then eliminate the record property names.
>
> The latter can be done crudely with

sed -i 's/:dummyroot\>//g' index
sed -i 's/:cmpfun\>//g' index

### TL;DR
Yay! It worked! And looking in the index file again, it's re-updated from
`:version "0.0.3"` to `:version 4`.

1. Can we close this issue, or do you want to keep it open to bug-hunt something
that is no longer an issue?

2. Perhaps this issue was triggerd by the fact that my elfeed version was old (weeks? or maybe months old?) when moving to emacs 26, and then upgrading to a recent elfeed _after_ moving to emacs 26?

**Comment by skeeto** [comment]
I still don't know where the :dummyroot property came from, but that 
explains the discrepancy with the mixture of old and new struct style.

Your elpa/ directory, including Elfeed, was compiled by Emacs 25, and it 
loaded under Emacs 26. This is one of the major issues with package.el. 
Byte-compiled lisp files generally aren't compatible between Emacs 
releases, but, despite this, package.el will silently load .elc files 
compiled under other releases (even future Emacs releases), leading to 
the sort of breakage we saw here. When you upgrade Emacs, the only safe 
way to use package.el is to purge your elpa/ directory and start fresh.

(Side note: To fix this, package.el should maintain separate elpa/ 
directories for each version of Emacs. It should also be a lot more 
careful about cross-package and cross-file macro contamination when 
updating packages, since it also gets this wrong.)

Emacs' byte code itself is backwards compatible — new versions of Emacs 
will execute old byte-code correctly — but macro expansions may no 
longer be valid. One of the major changes between Emacs 25 and 26 was a 
new fundamental type in the cl-defstruct macro. The sources compiled by 
Emacs 25 had been expanded to the original definition, not the Emacs 26 
definition, so you end up with the wrong behavior.

I'll close the issue since we mostly got to the bottom of it. There's 
not much I can do about bugs resulting from package.el mismanaging 
Elfeed builds.

**Comment by impaktor** [comment]
> Your elpa/ directory, including Elfeed, was compiled by Emacs 25, and it loaded under Emacs 26. This is one of the major issues with package.el. Byte-compiled lisp files generally aren't compatible between Emacs releases

Interesting. Is the incompatibility only between major versions?

> I'll close the issue since we mostly got to the bottom of it.

:+1: (I've removed the logs I linked to in my previous post).

Big thanks for helping me sort it out!

**Comment by skeeto** [comment]
As far as I know, Emacs doesn't use anything like semver (not that they 
should) and there are no promises that minor releases won't make big 
changes. For example, Emacs 24.3 deprecated cl and created a namespaced 
version as a new cl-lib package. Emacs 24.4 introduced new byte code 
opcodes. So, in practice, there can be incompatibilities for compiled 
lisp files even between minor releases. Emacs 26 is just a particular 
tricky case due to cl-defstruct changes.
